### ISSUE TITLE
A parse error has been introduced by #467

### ISSUE THREAD

**Issue by pocke** [whitequark/parser#473]
Since #467 (in other words, since v2.5.0.4), the parser raises a syntax error to the following code, but it is correct as Ruby code.

`test.rb`

```ruby
anything "#{[]}"
anything.anything do
  "#{anything}"
end
```

```bash
$ ruby-parse _2.5.0.4_ -E test.rb
anything "#{[]}"   
^~~~~~~~ tIDENTIFIER "anything"                 expr_cmdarg  [0 <= cond] [0 <= cmdarg] 
anything "#{[]}"   
         ^ tSTRING_BEG "\""                     expr_beg     [0 <= cond] [0 <= cmdarg] 
anything "#{[]}"   
          ^~ tSTRING_DBEG "\#{"                 expr_beg     [0 <= cond] [10 <= cmdarg] 
anything "#{[]}"   
            ^ tLBRACK "["                       expr_beg     [0 <= cond.1] [0 <= cmdarg.1] 
anything "#{[]}"   
             ^ tRBRACK "]"                      expr_end     [0 <= cond.1] [0 <= cmdarg.1] 
anything "#{[]}"   
              ^ tSTRING_DEND "}"                interp_string [0 <= cond.1] [0 <= cmdarg.1] 
anything "#{[]}"   
               ^ tSTRING_END "\""               expr_end     [0 <= cond] [10 <= cmdarg] 
anything "#{[]}"   
                ^ tNL nil                       line_begin   [0 <= cond] [10 <= cmdarg] 
anything.anything do   
^~~~~~~~ tIDENTIFIER "anything"                 expr_cmdarg  [0 <= cond] [1 <= cmdarg] 
anything.anything do   
        ^ tDOT "."                              expr_dot     [0 <= cond] [1 <= cmdarg] 
anything.anything do   
         ^~~~~~~~ tIDENTIFIER "anything"        expr_arg     [0 <= cond] [1 <= cmdarg] 
anything.anything do   
                  ^~ kDO_BLOCK "do"             expr_value   [0 <= cond] [1 <= cmdarg] 
test.rb:2:19: error: unexpected token kDO_BLOCK
test.rb:2: anything.anything do
test.rb:2:  
```

```bash
$ ruby-parse _2.5.0.3_ -E test.rb
anything "#{[]}"   
^~~~~~~~ tIDENTIFIER "anything"                 expr_cmdarg  [0 <= cond] [0 <= cmdarg] 
anything "#{[]}"   
         ^ tSTRING_BEG "\""                     expr_beg     [0 <= cond] [0 <= cmdarg] 
anything "#{[]}"   
          ^~ tSTRING_DBEG "\#{"                 expr_beg     [0 <= cond] [10 <= cmdarg] 
anything "#{[]}"   
            ^ tLBRACK "["                       expr_beg     [0 <= cond] [100 <= cmdarg] 
anything "#{[]}"   
             ^ tRBRACK "]"                      expr_end     [0 <= cond] [10 <= cmdarg] 
anything "#{[]}"   
              ^ tSTRING_DEND "}"                interp_string [0 <= cond] [10 <= cmdarg] 
anything "#{[]}"   
               ^ tSTRING_END "\""               expr_end     [0 <= cond] [1 <= cmdarg] 
anything "#{[]}"   
                ^ tNL nil                       line_begin   [0 <= cond] [1 <= cmdarg] 
anything.anything do   
^~~~~~~~ tIDENTIFIER "anything"                 expr_cmdarg  [0 <= cond] [0 <= cmdarg] 
anything.anything do   
        ^ tDOT "."                              expr_dot     [0 <= cond] [0 <= cmdarg] 
anything.anything do   
         ^~~~~~~~ tIDENTIFIER "anything"        expr_arg     [0 <= cond] [0 <= cmdarg] 
anything.anything do   
                  ^~ kDO "do"                   expr_value   [0 <= cond] [0 <= cmdarg] 
  "#{anything}"   
  ^ tSTRING_BEG "\""                            expr_cmdarg  [0 <= cond] [0 <= cmdarg] 
  "#{anything}"   
   ^~ tSTRING_DBEG "\#{"                        expr_cmdarg  [0 <= cond] [0 <= cmdarg] 
  "#{anything}"   
     ^~~~~~~~ tIDENTIFIER "anything"            expr_cmdarg  [0 <= cond] [0 <= cmdarg] 
  "#{anything}"   
             ^ tSTRING_DEND "}"                 interp_string [0 <= cond] [0 <= cmdarg] 
  "#{anything}"   
              ^ tSTRING_END "\""                expr_end     [0 <= cond] [0 <= cmdarg] 
  "#{anything}"   
               ^ tNL nil                        line_begin   [0 <= cond] [0 <= cmdarg] 
end   
^~~ kEND "end"                                  expr_end     [0 <= cond] [0 <= cmdarg] 
end   
   ^ tNL nil                                    line_begin   [0 <= cond] [0 <= cmdarg] 
   
^ false "$eof"                                  line_begin   [0 <= cond] [0 <= cmdarg] 
(begin
  (send nil :anything
    (dstr
      (begin
        (array))))
  (block
    (send
      (send nil :anything) :anything)
    (args)
    (dstr
      (begin
        (send nil :anything)))))
```

The syntax error is occurred with `--24` or `--25` option only. If `--23` option is received, it is not occurred.

Note: This problem was reported by @meganemura in https://github.com/rubocop-jp/issues/issues/40 (Japanese). Thank you!

**Comment by iliabylich** [comment]
@whitequark My initial assumption that `yychar` from bison can be replicated using `lexer.last_token` was wrong:
```
$ git diff
diff --git a/lib/parser/ruby25.y b/lib/parser/ruby25.y
index 157e393..b362e8d 100644
--- a/lib/parser/ruby25.y
+++ b/lib/parser/ruby25.y
@@ -873,6 +873,7 @@ rule
                       # For all other cases (like `m n` or `m n, []`) we simply put 1 to the stack
                       # and later lexer pushes corresponding bits on top of it.
                       last_token = @lexer.last_token[0]
+                      puts last_token
                       lookahead = last_token == :tLBRACK || last_token == :tLPAREN_ARG

                       if lookahead

$ rake generate && bin/ruby-parse --25 -E -e 'm "#{[]}"'
racc --superclass=Parser::Base lib/parser/ruby25.y -o lib/parser/ruby25.rb --no-line-convert
m "#{[]}"
^ tIDENTIFIER "m"                               expr_cmdarg  [0 <= cond] [0 <= cmdarg]
m "#{[]}"
  ^ tSTRING_BEG "\""                            expr_beg     [0 <= cond] [0 <= cmdarg]
tLBRACK
```

Initially I thought that I don't understand how LALR(1) parsers work 😄, but I think I understand it. Parser doesn't make another shift in this case (because it's not allowed, (1) means "no more than one shift to resolve conflicts", and in this case it was a shift to detect `command_args`), so `last_token` works exactly as `yychar` in bison. That's a lexer. It emits two tokens during one `advance` (we have a `@token_queue` that handles it).

So basically the general rule how to replicate the bug is: pass any code construction that emits two tokens during the first `advance` and pass it to the command call.

I guess `Lexer#advance` should store `@last_token`, that's the fix. @whitequark any thoughts about it? Does it sound right?
